ufl/referencevalue.py

```diff
@@ -39,7 +40,10 @@ def __init__(self, f):
 
     @property
     def ufl_shape(self):
-        return self.ufl_operands[0].ufl_element().reference_value_shape()
+        if not isinstance(self.ufl_operands[0], ExternalOperator):
+            return self.ufl_operands[0].ufl_element().reference_value_shape()
```

Comment:
This is a bad code smell. Why is the shape of an `ExternalOperator` not obtained in the same way as everything else?

Reply:
When differentiating an ExternalOperator its derivatives come out through the chain rule with increased shapes. Now these external operator's derivatives are Coefficient whose the shape are different from the one associated to their function spaces, which is logically problematic for referencing since concretely Coefficients have not been designed to have a shape different from Coefficient.ufl_element().reference_value_shape(). If we think of other cases where a shape changing occurs, like when apply a differential operator on a function (e.g Grad), in these cases there is a mechanism to handle the referencing (e.g ReferenceGrad/ReferenceDiv). 

That explains why we introduce the "original_function_space" and also why this exception exists.